append-etl/etl-append-data-to-tables.scala

```diff
@@ -0,0 +1,127 @@
+val dataAppendWithSvkeys = dataAppend
+  .withColumn("trueSvkeys", splitSvkeys(col("ATTRIBUTES")))
+  .cache()
+
+// Trigger an action to cache the DataFrame
+dataAppendWithSvkeys.count()
+
+// Narrow the taxonomic scope by extracting distinct svkeys
+val superSetSvkeys = dataAppendWithSvkeys
+  .select(explode(col("trueSvkeys")).as("svkey"))
+  .distinct()
+  .as[String]
+  .collect()
+
+// Filter taxonomy based on the relevant svkeys and create 'attributeString'
+val focalTaxonomy = dataAppendTaxonomy
+  .where(col("survey_value_key").isin(superSetSvkeys: _*))
+  .withColumn("attributeString", concat_ws("_", col("AttributeKey"), col("survey_value_key")))
+
+// Create a map for single select attributes
+val singleSelectColumnMap: Map[String, String] = focalTaxonomy
+  .filter(col("Attribute Type") === "Single Select")
+  .select("survey_value_key", "AttributeKey")
+  .distinct()
+  .as[(Int, String)]
+  .collect()
+  .map { case (svkey, attrKey) => svkey.toString -> attrKey.toString }
+  .toMap
+
+
+// Extract distinct single and multi-select attributes
+val singleSelectAttributes = singleSelectColumnMap.values.toArray.distinct
+val multiSelectAttributes = focalTaxonomy
+  .filter(col("Attribute Type") === "Multi Select")
+  .select("attributeString")
+  .distinct()
+  .as[String]
+  .collect()
+
+val allAttributes = focalTaxonomy
+  .select("attributeString")
+  .distinct()
+  .as[String]
+  .collect()
+
+// Broadcast the singleSelectColumnMap for efficient lookup in UDF
+val broadcastMap = spark.sparkContext.broadcast(singleSelectColumnMap)
+// UDF to map svkeys to a single select attribute
+val mapSvkeyToAttr = udf((trueSvkeys: Seq[String], attr: String) => {
+  // Access the broadcasted map within the UDF's closure
+  trueSvkeys
+    .flatMap(svkey => broadcastMap.value.get(svkey))
+    .find(_ == attr)
+    .getOrElse(null)
+})
+
+// Create DataFrame for single selects by adding a new column for each single select attribute
+val singleSelectDF = singleSelectAttributes.foldLeft(dataAppendWithSvkeys) { (df, attr) =>
+  df.withColumn(
+    attr,
+    mapSvkeyToAttr(col("trueSvkeys"), lit(attr))
+  )
+}.drop("trueSvkeys", "ATTRIBUTES")
+
+// Process multi-select attributes
+val multiSelectDF = dataAppendWithSvkeys
+  .select(col("ID"), explode(col("trueSvkeys")).as("survey_value_key"))
+  .join(focalTaxonomy, "survey_value_key")
+  .groupBy("ID")
+  .pivot("attributeString", multiSelectAttributes)
+  .agg(count("survey_value_key"))
+  .na.fill(0)
+
+// Alternatively, process all attributes as multi-selects
+val allAsMultiSelectDF = dataAppendWithSvkeys
+  .select(col("ID"), explode(col("trueSvkeys")).as("survey_value_key"))
+  .join(focalTaxonomy, "survey_value_key")
+  .groupBy("ID")
+  .pivot("attributeString", allAttributes)
+  .agg(count("survey_value_key"))
+  .na.fill(0)
+
+// Save the processed DataFrames
+singleSelectDF.write.mode("overwrite").save(params.outDir + "singleSelectDF")
```

Comment:
In my testing, it looked like the values for the attribute key columns in this dataset were attribute keys not survey value keys.

Reply:
I've fixed this